src/runtime_src/core/tools/xbmgmt2/OO_UpdateBase.cpp

```diff
@@ -0,0 +1,752 @@
+    // Print progress while sc is flashed
+    t = std::thread([&progress_reporter, &done]() {
+      unsigned int counter = 0;
+      while ((counter < progress_reporter.get()->getMaxIterations()) && !done) {
+        progress_reporter.get()->update(counter++);
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+      }
+    });
+    xrt_core::device_update<xrt_core::query::program_sc>(dev.get(), val);
+    done = true;
+    progress_reporter.get()->finish(true, "SC firmware image has been programmed successfully.");
+    t.join();
+  }
+  catch (const xrt_core::query::sysfs_error& e) {
+    done = true;
+    progress_reporter.get()->finish(false, "Failed to update SC flash image.");
+    t.join();
+    throw xrt_core::error(std::string("Error accessing sysfs entry : ") + e.what());
+  }
+}
+
+// Update SC firmware on the board
+static void
+update_SC(unsigned int  index, const std::string& file)
+{
+  Flasher flasher(index);
+
+  if (!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % index));
+
+  auto dev = xrt_core::get_mgmtpf_device(index);
+
+  bool is_versal = xrt_core::device_query<xrt_core::query::is_versal>(dev);
+  if (is_versal) {
+    update_versal_SC(dev);
+    return;
+  }
+
+  //if factory image, update SC
+  auto is_mfg = xrt_core::device_query<xrt_core::query::is_mfg>(dev);
+  if (is_mfg) {
+    std::unique_ptr<firmwareImage> bmc = std::make_unique<firmwareImage>(file, BMC_FIRMWARE);
+    if (bmc->fail())
+      throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file));
+
+    if (flasher.upgradeBMCFirmware(bmc.get()) != 0)
+      throw xrt_core::error("Failed to update SC flash image");
+    return;
+  }
+
+  // If SC is fixed, stop flashing immediately
+  if (is_SC_fixed(index))
+    throw xrt_core::error("SC is fixed, unable to flash image.");
+
+  // Mgmt pf needs to shutdown so that the board doesn't brick
+  try {
+    dev->device_shutdown();
+  }
+  catch (const xrt_core::error& e) {
+    throw xrt_core::error(std::string("Only proceed with SC update if all user applications for the target card(s) are stopped. ") + e.what());
+  }
+
+  std::unique_ptr<firmwareImage> bmc = std::make_unique<firmwareImage>(file, BMC_FIRMWARE);
+
+  if (bmc->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file));
+
+  if (flasher.upgradeBMCFirmware(bmc.get()) != 0)
+    throw xrt_core::error("Failed to update SC flash image");
+
+  // Bring back mgmt pf
+  try {
+    dev->device_online();
+  }
+  catch (const xrt_core::error& e) {
+    throw xrt_core::error(e.what() + std::string(" Please warm reboot."));
+  }
+
+  std::cout << boost::format("%-8s : %s \n\n") % "INFO" % "SC firmware image has been programmed successfully.";
+}
+
+// Helper function for header info
+static std::string
+file_size(const std::string & _file)
+{
+  std::ifstream in(_file.c_str(), std::ifstream::ate | std::ifstream::binary);
```

Comment:
```suggestion
  std::ifstream in(_file, std::ifstream::ate | std::ifstream::binary);
```
Do you have a C-string obsession? :-)